### DISCUSSION TITLE
[Axum0.7x]how to use into_make_service_with_connect_info in low-level-rustls

### DISCUSSION THREAD

**Discussion by yuhaya** [tokio-rs/axum#2397] [Q&A] [answered]
```rust
async fn https_server(ports: Ports) {
    let app = server_routes();
    let addr = format!("{}:{}", env::var("HOST").unwrap(), ports.https);

    println!("listening on {}", addr);

    let rustls_config = rustls_server_config(
        PathBuf::from(env::var("HTTPS_KEY").unwrap()),
        PathBuf::from(env::var("HTTPS_CERT").unwrap()),
    );

    let tls_acceptor = TlsAcceptor::from(rustls_config);
    let tcp_listener = TcpListener::bind(&addr).await.unwrap();

    pin_mut!(tcp_listener);
    loop {
        let tower_service = app
            .clone()
            .into_make_service_with_connect_info::<SocketAddr>();
        let tls_acceptor = tls_acceptor.clone();

        // Wait for new tcp connection
        let (cnx, addr) = tcp_listener.accept().await.unwrap();

        tokio::spawn(async move {
            // Wait for tls handshake to happen
            let Ok(stream) = tls_acceptor.accept(cnx).await else {
                error!("error during tls handshake connection from {}", addr);
                return;
            };

            let stream = TokioIo::new(stream);
            let hyper_service = hyper::service::service_fn(move |request: Request<Incoming>| {
                tower_service.clone().call(request)
            });

            let ret = hyper_util::server::conn::auto::Builder::new(TokioExecutor::new())
                .serve_connection_with_upgrades(stream, hyper_service)
                .await;

            if let Err(err) = ret {
                warn!("error serving connection from {}: {}", addr, err);
            }
        });
    }
}
```

### Summary
```rust
tower_service.clone().call(request)
```
### Error Notice
> the trait bound `std::net::SocketAddr: axum::extract::connect_info::Connected<hyper::Request<hyper::body::Incoming>>` is not satisfied
the following other types implement trait `axum::extract::connect_info::Connected<T>`:
  <std::net::SocketAddr as axum::extract::connect_info::Connected<axum::serve::IncomingStream<'_>>>
  <std::net::SocketAddr as axum::extract::connect_info::Connected<std::net::SocketAddr>>
required for `axum::extract::connect_info::IntoMakeServiceWithConnectInfo<axum::Router, std::net::SocketAddr>` to implement `tower::Service<hyper::Request<hyper::body::Incoming>>`

```rust
let ret = hyper_util::server::conn::auto::Builder::new(TokioExecutor::new())
                .serve_connection_with_upgrades(stream, hyper_service)
                .await;
```
### Error Notice
>  expected `ResponseFuture<Router, SocketAddr>` to be a future that resolves to `Result<Response<_>, _>`, but it resolves to `Result<AddExtension<Router, ConnectInfo<SocketAddr>>, Infallible>`
expected enum `std::result::Result<axum::middleware::AddExtension<axum::Router, axum::extract::ConnectInfo<std::net::SocketAddr>>, std::convert::Infallible>`
   found enum `std::result::Result<hyper::Response<_>, _>`
required for `hyper::service::util::ServiceFn<{closure@server/src/lib.rs:120:60: 120:93}, hyper::body::Incoming>` to implement `hyper::service::Service<hyper::Request<hyper::body::Incoming>>`

> expected `ResponseFuture<Router, SocketAddr>` to be a future that resolves to `Result<Response<_>, _>`, but it resolves to `Result<AddExtension<Router, ConnectInfo<SocketAddr>>, Infallible>`
expected enum `std::result::Result<hyper::Response<_>, _>`
   found enum `std::result::Result<axum::middleware::AddExtension<axum::Router, axum::extract::ConnectInfo<std::net::SocketAddr>>, std::convert::Infallible>`
required for `hyper::service::util::ServiceFn<{closure@server/src/lib.rs:120:60: 120:93}, hyper::body::Incoming>` to implement `hyper::service::Service<hyper::Request<hyper::body::Incoming>>`
### axum version
0.71

**Comment by davidpdrsn** [comment]
Please post your code and fix the formatting.

**Reply by yuhaya** [reply]
ok,Thanks for your help, the code has been added and the format has been corrected.

**Comment by yuhaya** [comment, marked as answer]
[We don't need to call `poll_ready` because `IntoMakeServiceWithConnectInfo` is always ready.](https://github.com/tokio-rs/axum/blob/b02ce307371a973039018a13fa012af14775948c/examples/serve-with-hyper/src/main.rs#L98)
